dronegen/misc.go

```diff
@@ -14,9 +14,185 @@
 
 package main
 
+import (
+	"fmt"
+	"strings"
+)
+
 func promoteBuildPipeline() pipeline {
 	// TODO: migrate
-	return pipeline{}
+
+	aptPipeline := promoteAptPipeline()
+	return aptPipeline
+}
+
+// This function calls the build-apt-repos tool which handles the APT portion of RFD 0058.
+func promoteAptPipeline() pipeline {
+	testVersion := "6.2.31"
+	aptVolumeName := "aptrepo"
+	artifactPath := "/go/artifacts"
+
+	p := newKubePipeline("publish-apt-new-repos")
+	// p.Trigger = triggerPromote
+	p.Trigger = trigger{
+		Event:  triggerRef{Include: []string{"push"}},
+		Branch: triggerRef{Include: []string{"rfd/0058-package-distribution"}},
+	}
+	p.Workspace = workspace{Path: "/go"}
+	p.Volumes = []volume{
+		{
+			Name: aptVolumeName,
+			Claim: &volumeClaim{
+				Name: "drone-s3-aptrepo-pvc",
+			},
+		},
+		volumeTmpfs,
+	}
+	p.Steps = []step{
+		// {
+		// 	Name:  "Verify build is tagged",
+		// 	Image: "alpine:latest",
+		// 	Commands: []string{
+		// 		"[ -n ${DRONE_TAG} ] || (echo 'DRONE_TAG is not set. Is the commit tagged?' && exit 1)",
+		// 	},
+		// },
+		{
+			Name:  "Check out code",
+			Image: "alpine/git:latest",
+			Environment: map[string]value{
+				"GITHUB_PRIVATE_KEY": {
+					fromSecret: "GITHUB_PRIVATE_KEY",
+				},
+			},
+			Commands: aptToolCheckoutCommands(),
+		},
+		{
+			Name:  "Download artifacts",
+			Image: "amazon/aws-cli",
+			Environment: map[string]value{
+				"AWS_S3_BUCKET": {
+					fromSecret: "AWS_S3_BUCKET",
+				},
+				"AWS_ACCESS_KEY_ID": {
+					fromSecret: "AWS_ACCESS_KEY_ID",
+				},
+				"AWS_SECRET_ACCESS_KEY": {
+					fromSecret: "AWS_SECRET_ACCESS_KEY",
+				},
+				"ARTIFACT_PATH": {
+					raw: artifactPath,
+				},
+			},
+			Commands: []string{
+				"mkdir -pv \"$ARTIFACT_PATH\"",
+				strings.Join(
+					[]string{
+						"aws s3 sync",
+						"--no-progress",
+						"--delete",
+						"--exclude \"*\"",
+						"--include \"*.deb*\"",
+						// TODO change from sprintf to `s3://$AWS_S3_BUCKET/teleport/tag/${DRONE_TAG##v}/` after testing
+						fmt.Sprintf("s3://$AWS_S3_BUCKET/teleport/tag/%s/", testVersion),
+						"\"$ARTIFACT_PATH\"",
+					},
+					" ",
+				),
+			},
+		},
+		{
+			Name: "Publish debs to APT repos",
+			// TODO set this if `step` support s https://docs.drone.io/pipeline/ssh/syntax/parallelism/ in the future
+			// DependsOn: []string {
+			// 	"Check out code",
+			// 	"Download artifacts",
+			// },
+			Image: "golang:1.18.1-bullseye",
+			Environment: map[string]value{
+				"APT_S3_BUCKET": {
+					fromSecret: "APT_REPO_NEW_AWS_S3_BUCKET",
+				},
+				"AWS_REGION": {
+					raw: "us-west-2",
+				},
+				"AWS_ACCESS_KEY_ID": {
+					fromSecret: "APT_REPO_NEW_AWS_ACCESS_KEY_ID",
+				},
+				"AWS_SECRET_ACCESS_KEY": {
+					fromSecret: "APT_REPO_NEW_AWS_SECRET_ACCESS_KEY",
+				},
+				"ARTIFACT_PATH": {
+					raw: artifactPath,
+				},
+				"GNUPGHOME": {
+					raw: "/tmpfs/gnupg",
```

Comment:
Can simultaneous builds stomp on this directory?

Reply:
Yes they can. It's important to make sure only one migration or promotion pipeline is running at a time.

Reply:
Sorry yes that is correct. I misread which line the comment was referring to.